stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift

```diff
@@ -64,12 +67,38 @@ func posixPipe() -> (readFD: CInt, writeFD: CInt) {
 /// stderr.
 public func spawnChild(args: [String])
   -> (pid: pid_t, stdinFD: CInt, stdoutFD: CInt, stderrFD: CInt) {
+  let childStdout = posixPipe()
+  let childStdin = posixPipe()
+  let childStderr = posixPipe()
+
+#if os(Android)
+  // posix_spawn isn't available on Android. Instead, we fork and exec.
+  let pid = fork()
+  if pid < 0 {
+    preconditionFailure("fork() failed")
+  } else if pid == 0 {
+    // pid of 0 means we are now in the child process.
+    // Capture the output and execute the program.
+    dup2(childStdout.writeFD, STDOUT_FILENO)
+    dup2(childStdin.readFD, STDIN_FILENO)
+    dup2(childStderr.writeFD, STDERR_FILENO)
+
+    var result: Int32 = 0
+    withArrayOfCStrings([Process.arguments[0]] + args) {
+      result = execve(Process.arguments[0], $0, _getEnviron())
+    }
+
+    if result != 0 {
+      preconditionFailure("execve() failed. errno: \(errno)")
+    }
+    _exit(126)
```

Comment:
Don't worry about the wait, great work on this PR! So the reason we need `child2parent` is to communicate from child to parent.

There's basically two possible outcomes

1. everything went fine
2. there was an error in the child (for example file to execute not found, no permission or what not)

In both cases, it's crucial that the child can tell the parent. The easiest way on UNIX is to just open a pipe (`child2parent`) in the parent before the fork. Because file descriptors are forked as well, we will have a communication channel from the child to the parent after the fork. `child2parent[0]` is the reading end, so that's the file descriptor the parent keeps open. And `child2parent[1]` is the writing end, that's the fd the child keeps open. In other words, if the child writes something to `child2parent[1]`, the parent will receive it on `child2parent[0]`. So far so good, now we have a communication facility.

If there was an error `execve`ing in the child, we can just write the `errno` value that it got into `child2parent[1]` and the parent will know what exactly went wrong. For example if the file to be executed wasn't found, a `ENOENT` will be transmitted and the parent can fail appropriately.

The seemingly easier case is actually the harder: What to do if there was _no_ error in the child, i.e. the child executed another process. Then `execve` succeeds and we replaced the process with some other binary. The file descriptor `child2parent[1]` will now still be open so we could put something in there which tells the parent that everything went fine. However, the program we executed probably doesn't know it's supposed to do that. The standard trick to get around this issue is to set the file descriptor `child2parent[1]` as `FD_CLOEXEC`. If `FD_CLOEXEC` is set, the file descriptor is closed _by the OS kernel_ automatically if `execve` succeeds. That's perfect for us: If `child2parent[1]` is closed without anything ever being written to the pipe, the parent can now learn that `execve` has succeeded which is exactly the missing big.

So the logic in the parent will be

1. try to read on `parent2child[0]`
2a. if read returns `0` (meaning EOF), the parent knows everything went fine, the child successfully executed another program
2b. if read returns some bytes, we know that something went wrong and the value that we read is the `errno` of the child that it got when trying to execute the other program. That's awesome as we now know in the parent what went wrong in the child :).

Hope this makes any sense, if not or if there are any questions left, feel free to ask.

1.